### ISSUE TITLE
"Error: NGCC failed" when using `"somepackage": "file:app/somepackage"` in package.json

### ISSUE THREAD

**Issue by kmaraz** [angular/angular#38832]
# 🐞 bug report

### Affected Package
@ngtools/webpack@10.1.0
@angular/compiler@10.1.0
@angular/compiler-cli@10.1.0

Now I have version 10.1.1, but bug is still there.

### Is this a regression?

Yes, the previous version in which this bug was not present was: 10.0.14

### Description

I am using custom Webpack configuration as I am coming from NgUpgrade world. I am using in my work project multiple git submodules. I am linking their package.json via `"somepackage": "file:app/somepackage"` format (see in my github repo package.json).

## 🔬 Minimal Reproduction

https://github.com/kmaraz/ngcc-failed

Just install packages via `yarn` and then build project with Ivy: `yarn run failsdev`.
Note: when you remove `somepackage` from package.json, run yarn, then everything works as expected.

## 🔥 Exception or Error
100% times build just hangs, sometimes when it hangs and I terminate it, I get Error: NGCC failed.

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular packages in version 10.1.1
</code></pre>

**Anything else relevant?**
I am using Windows Subsystem for Linux.

**Comment by JoostK** [comment]
Hi, using your exact repro steps I don't get any errors; the app builds successfully in Ivy. I'm on macOS, so that may make a difference here.

**Comment by gkalpak** [comment]
Maybe it is somehow related to WSL 😱
@kmaraz, have you cloned the repo on the Windows filesystem or in WSL?
Also, what version of WSL are you using and what linux distro/version?

**Comment by kmaraz** [comment]
I have cloned this repo in WSL2. I have Ubuntu 20.04.1 LTS.

**Comment by kmaraz** [comment]
This is the error:
 ```
Error: NGCC failed.
    at NgccProcessor.process (/mnt/e/projects/slido-admin/node_modules/@ngtools/webpack/src/ngcc_processor.js:119:19)
    at /mnt/e/projects/slido-admin/node_modules/@ngtools/webpack/src/angular_compiler_plugin.js:590:31
    at SyncHook.eval [as call] (eval at create (/mnt/e/projects/slido-admin/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:9:1)
    at SyncHook.lazyCompileHook (/mnt/e/projects/slido-admin/node_modules/tapable/lib/Hook.js:154:20)
    at webpack (/mnt/e/projects/slido-admin/node_modules/webpack/lib/webpack.js:55:30)
    at startDevServer (/mnt/e/projects/slido-admin/node_modules/webpack-dev-server/bin/webpack-dev-server.js:94:16)
    at /mnt/e/projects/slido-admin/node_modules/webpack-dev-server/bin/webpack-dev-server.js:166:3
    at /mnt/e/projects/slido-admin/node_modules/webpack-dev-server/lib/utils/processOptions.js:33:9
    at processTicksAndRejections (internal/process/task_queues.js:93:5)```

**Comment by gkalpak** [comment]
I tried this with Ubuntu 20.04.1 in WSL2 on my machine and it seems to work fine. It sounds like something that might be specific to your setup, @kmaraz 😞

What happens if you run `yarn` and then `yarn ngcc --create-ivy-entry-points --first-only --loglevel debug --properties module`?

**Comment by kmaraz** [comment]
I am getting this error:
```
/node_modules/.bin/ngcc --create-ivy-entry-points --first-only --loglevel debug --properties
Error: No supported format property to consider among []. Supported properties: fesm2015, fesm5, es2015, esm2015, esm5, main, module, browser
    at ensureSupportedProperties (/mnt/e/projects/slido-admin/node_modules/@angular/compiler-cli/ngcc/src/main.js:102:19)
    at Object.mainNgcc (/mnt/e/projects/slido-admin/node_modules/@angular/compiler-cli/ngcc/src/main.js:55:45)
    at /mnt/e/projects/slido-admin/node_modules/@angular/compiler-cli/ngcc/main-ngcc.js:34:53
    at step (/mnt/e/projects/slido-admin/node_modules/@angular/compiler-cli/node_modules/tslib/tslib.js:140:27)
    at Object.next (/mnt/e/projects/slido-admin/node_modules/@angular/compiler-cli/node_modules/tslib/tslib.js:121:57)
    at /mnt/e/projects/slido-admin/node_modules/@angular/compiler-cli/node_modules/tslib/tslib.js:114:75
    at new Promise (<anonymous>)
    at Object.__awaiter (/mnt/e/projects/slido-admin/node_modules/@angular/compiler-cli/node_modules/tslib/tslib.js:110:16)
    at /mnt/e/projects/slido-admin/node_modules/@angular/compiler-cli/ngcc/main-ngcc.js:28:39
    at /mnt/e/projects/slido-admin/node_modules/@angular/compiler-cli/ngcc/main-ngcc.js:51:15
error Command failed with exit code 1.
```

**Comment by kmaraz** [comment]
After that everything looks fine. But it started happening in my other project. I don't know why it doesn't happen anymore, neither do I know why it started to happen in my other project. Sadly I can't share its sources. I wish there was some error message or someting to point me somewhere, but nothing. Anyway, thanks for help. I'll try to come up with more replicable repo.

**Comment by gkalpak** [comment]
OOC, what running `yarn ngcc --create-ivy-entry-points --first-only --loglevel debug --properties module` on the other project fix it too?

**Comment by kmaraz** [comment]
Nope, it didn't. Also Now I tested the second project on Window 10 via command line (no WSL, bash, etc.) and I've got the same error. So It must be someting related to that project. I'll try to pinpoint by gradual uprading of the packages at least to the latest stable version and then I'll try to compare diffs in the Angular code. Maybe I'll come up with something more concrete.

**Comment by kmaraz** [comment]
I was able to pinpoint the problem to this block of code:

```
 // We spawn instead of using the API because:
        // - NGCC Async uses clustering which is problematic when used via the API which means
        // that we cannot setup multiple cluster masters with different options.
        // - We will not be able to have concurrent builds otherwise Ex: App-Shell,
        // as NGCC will create a lock file for both builds and it will cause builds to fails.
        const { status, error } = child_process_1.spawnSync(process.execPath, [
            require.resolve('@angular/compiler-cli/ngcc/main-ngcc.js'),
            '--source',
            this._nodeModulesDirectory,
            '--properties',
            ...this.propertiesToConsider,
            '--first-only',
            '--create-ivy-entry-points',
            '--async',
            '--tsconfig',
            this.tsConfigPath,
            '--use-program-dependencies',
        ], {
            stdio: ['inherit', process.stderr, process.stderr],
        });
```
Especially this line:
```
            '--use-program-dependencies',
```
When I comment it out, everything works fine, not matter what versions (or combination) of packages I use. But when this is turned on, webpack build just hangs and then I get the error provided [here.](https://github.com/angular/angular/issues/38832#issuecomment-692139640)

This code is present in `@ngtools/webpack` from the version `10.1.0-next.0`.

So I am pretty sure `--use-program-dependencies` is the "problem", but not sure why as I didn't study the sources of `main-ngcc.js` yet.

**Comment by petebacondarwin** [comment]
`--use-program-dependencies` will parse the imports in your project to work out what packages are actually being used. Based on that it will then determine which libraries/entry-points need processing.

I note that in your tsconfig.json you are using `exclude` to specify which files to load.
I suspect that the package.json/tsconfig.json in the `app/somepackage` folder is being picked up by TS as something to compile but the `main` file in that package does not exist.

Can you try running ngcc directly but including that `--use-program-dependencies` option. This may give a more specific error.

**Comment by kmaraz** [comment]
`yarn ngcc --use-program-dependencies  --loglevel debug` hangs on:
```
The lock-file path is /mnt/e/projects/app/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__
```
After I try to kill the process I will get:
```
 Debug: Attempting to remove lock-file at /mnt/e/projects/app/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.
Debug: PIDs match (23327), so removing /mnt/e/projects/app/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.
```
```

**Comment by petebacondarwin** [comment]
Removing the exclude would still cause TS to include the `app/somepackage` I believe. Try specifying:

```
  "files": ["app/index.ts"]
```

**Comment by kmaraz** [comment]
Looks like it still hangs on the last line:
```
yarn ngcc --use-program-dependencies --loglevel debug
yarn run v1.22.4
$ /mnt/e/projects/app/node_modules/.bin/ngcc --use-program-dependencies --loglevel debug
Debug: Forking unlocker child-process
Debug: Attemping to write lock-file at /mnt/e/projects/app/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__ with PID 23405
Debug: Written lock-file at /mnt/e/projects/app/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__ with PID 23405
Debug: Running ngcc on ClusterExecutor (using 5 worker processes).
Debug: Analyzing entry-points...
Debug: Using the program from /mnt/e/projects/app/tsconfig.json to seed the entry-point finding.
Debug: Collecting dependencies from the following files:
- /mnt/e/projects/app/app/index.ts
Debug: Starting unlocker at process 23412 on behalf of process 23405
Debug: The lock-file path is /mnt/e/projects/app/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__ 
```

**Comment by petebacondarwin** [comment]
Interesting... so the last two lines are from a separate "unlocker process", which runs independently, so these can be ignored.

The place it is actually hanging is 

```
Debug: Using the program from /mnt/e/projects/app/tsconfig.json to seed the entry-point finding.
Debug: Collecting dependencies from the following files:
- /mnt/e/projects/app/app/index.ts
```

This should find the two libraries `@angular/core` and `@angular/platform-browser-dynamic` but this appears to be hanging.

I note that the process of finding dependencies is recursive. I wonder if the way your system is setup there is an infinite recursion going on where it gets into an import cycle??

**Comment by petebacondarwin** [comment]
It might be worth, now, trying to debug into the ngcc process:

```
node --inspect-brk node_modules/.bin/ngcc --use-program-dependencies --loglevel debug
```

Then attach a debugger to this (perhaps VS Code or Chrome) and setup into the `getInitialEntryPointPaths()` function here: https://github.com/angular/angular/blob/master/packages/compiler-cli/ngcc/src/entry_point_finder/program_based_entry_point_finder.ts#L47

**Comment by kmaraz** [comment]
So I tried to find what's happening inside the compiler and I found out that essentially `node node_modules/.bin/ngcc --use-program-dependencies --loglevel debug` works perfectly and more importantly, after that I can run my Webpack dev server and it just works.

But...
On the Windows 10 command line setup it took: `Debug: Run ngcc in 397s.`
On the same machine, but through WSL it took: `Debug: Run ngcc in 6800s.` which is quite a lot.

**Comment by gkalpak** [comment]
Looking at your previous comments, it looks like you are using WSL2 to access files on thr Windows filesystem (on `E:\projects\app\`). WSL2 is known for being quite slow at accessing the Windows filesystem. See https://github.com/microsoft/WSL/issues/4197#issuecomment-604592340 for a more detailed explanation.

In fact, if you plan to use WSL on files stored on the Windows filesystem, it is recommended in the [official docs][1] to stick to WSL1. If you use WSL2 it is [recommended][2] to only work with files on the Linux filesystem.

##
Based on the above, the times you are seeing are not that surprizing and there is probably not much we can do on our side (for reducing the perf difference on WSL vs Windows when accessing the Windows filesystem).

What are the times when you clone the repo in your Linux filesystem?

[1]: https://docs.microsoft.com/en-us/windows/wsl/compare-versions#exceptions-for-using-wsl-1-rather-than-wsl-2
[2]: https://docs.microsoft.com/en-us/windows/wsl/compare-versions#performance-across-os-file-systems

**Comment by kmaraz** [comment]
Oh, wow! So, based on what you said @gkalpak, I have moved my project to native WSL2 home directory and NGCC time is now down to 30seconds :-O. This is a **completely different** world to me now.

I am really grateful to all of you guys with helping me to debug my problem. I have learned multiple things, like debugging nodejs projects, or about filesystems and how they cooperate.

So let me thank you again!
